reinitialize-pods/src/lib.rs

```diff
@@ -0,0 +1,176 @@
+    tokio::pin!(pod_evts);
+    while let Some(evt) = pod_evts.next().await {
+        tracing::trace!(?evt);
+        if let watcher::Event::Applied(pod) = evt {
+            let status = if let Some(ref status) = pod.status {
+                status.clone()
+            } else {
+                tracing::info!("Skipped, no status");
+                continue;
+            };
+
+            let terminated = status
+                .init_container_statuses
+                .unwrap_or_default()
+                .iter()
+                .find(|status| status.name == "linkerd-network-validator")
+                .and_then(|status| status.last_state.as_ref())
+                .and_then(|status| status.terminated.as_ref())
+                .map(|term| term.exit_code == UNSUCCESSFUL_EXIT_CODE)
+                .unwrap_or(false);
+
+            let evicted = status
+                .conditions
+                .as_ref()
+                .and_then(|conds| {
+                    conds.iter().find(|cond| {
+                        cond.reason
+                            .as_ref()
+                            .is_some_and(|reason| reason == CONDITION_EVICTED_REASON)
+                    })
+                })
+                .is_some();
+
+            if terminated && !evicted {
+                let namespace = pod.namespace().unwrap();
+                let name = pod.name_any();
+                // this avoids blocking the event loop
+                match tx.try_send(pod.clone()) {
+                    Ok(_) => {}
+                    Err(TrySendError::Full(_)) => {
+                        tracing::warn!(
+                            name = format!("{namespace}/{name}"),
+                            "Dropped event (channel full)"
+                        );
+                        metrics.queue_overflow.inc();
+                    }
+                    Err(TrySendError::Closed(_)) => tracing::warn!(
+                        name = format!("{namespace}/{name}"),
+                        "Dropped event (channel closed or dropped)"
+                    ),
+                }
+            }
+        }
+    }
+}
+
+async fn process_pods(
+    client: Client,
+    controller_pod_name: String,
+    mut rx: Receiver<Pod>,
+    metrics: Metrics,
+) {
+    while let Some(pod) = rx.recv().await {
+        let namespace = pod.namespace().unwrap();
+        let name = pod.name_any();
+        let pods = kube::Api::<Pod>::namespaced(client.clone(), &namespace);
+        let evict_res = pods.evict(&name, &Default::default()).await;
+        match evict_res {
+            Ok(_) => {
+                tracing::info!(name = format!("{namespace}/{name}"), "Evicting pod");
+                metrics.evicted_pods.inc();
+                if let Err(err) =
+                    publish_k8s_event(client.clone(), controller_pod_name.clone(), &pod).await
+                {
+                    tracing::warn!(%err, name = format!("{namespace}/{name}"), "Error publishing event");
```

Comment:
nit: to avoid multiple string allocations, we could just declare it once as `let pod_id = format!("{namespace}/{name}");`. Or alternatively, we could do what we do in the policy controller `tracing::warn!(%err, %namespace, %name, "My msg")`.

Reply:
Makes sense :+1: